src/vm/methodtable.cpp

```diff
@@ -1709,6 +1714,13 @@ TypeHandle::CastResult MethodTable::CanCastToClassNoGC(MethodTable *pTargetMT)
     }
     CONTRACTL_END
 
+    // allow an object of type T to be cast to Nullable<T> (they have the same representation)
+    if (pTargetMT->IsNullable() && 
+        pTargetMT->GetInstantiation()[0] == TypeHandle(this))
+    {
+        return TypeHandle::CanCast;
+    }
+
```

Comment:
This check is checking for identical matches in the Nullable, but the check above in CanCastToClass is performing an equivalence check. 

However, in this function, an equivalent, but not identical struct type will return CannotCast instead of MaybeCast as it should.

Reply:
I think the general `HasTypeEquivalence()` check below will make us to return `MaybeCast` 

---
In reply to: [306598756](https://github.com/dotnet/coreclr/pull/23548#discussion_r306598756) [](ancestors = 306598756)

Reply:
Yes. If either type has `HasTypeEquivalence()`, we will return `MaybeCast`. 
We do that check later. I assume because equivalency is less common than nullable or variance, which we check before it.
 

---
In reply to: [306996204](https://github.com/dotnet/coreclr/pull/23548#discussion_r306996204) [](ancestors = 306996204,306598756)